### PR TITLE
Clarify get_pilot_desired_lean_angles and fix limit

### PR DESCRIPTION
**Pull request by lthall**
I apologise for having another go at this function so quickly. While the equation in the previous PR was correct the method the maximum lean angle was limited was not. The previous code resulted in the following shape:

image

The curves are nearly correct but the maximum lean angle is reduced on the diagonals.

The changes below simplify and clarify the equations further by using vector2f and now handle the limits correctly. The new limits look like this:
 
image

Credit to Mark Whitehorn for discovering the limit problem.

### CODE REVIEW

**Comment by kd0aij** [comment]
We almost agree now. There's a phase difference which is at a maximum here (red is your patch, yellow is mine, green is the stick)
This is looking straight down the Z axis. I updated the VPython web code if you'd like to look at the animation.
image

**Comment by lthall** [comment]
I think Mark and I feel good about this now.

**Review by tridge — CHANGES_REQUESTED** [review]
i think this should be in control.cpp, and then someone should write a test

**Comment by tridge** [comment]
the change is also needed in quadplanes

**Comment by rmackay9** [comment]
if we move this to a common function then we should probably change the inputs of the function be in standard units (e.g. degrees instead of centi-degrees)

**Comment by kd0aij** [comment]
@lthall  See if this version of my test code works: 
https://glowscript.org/#/user/kd0aij/folder/ArduCopter/program/stickmaptest.py
Red is my python implementation of your code, green is my alternate.

**Comment by kd0aij** [comment]
@lthall This python program is an independent implementation of the rc input to roll/pitch conversion.  
https://glowscript.org/#/user/kd0aij/folder/ArduCopter/program/EulerRotations
Let me know if the roll and pitch sliders do what you expect them to in all orientations.  If it does, it can serve as the basis of a test.  If not, it should help us figure out where we have different expectations.

**Comment by lthall** [comment]
Added Test from Mark.

File: ArduCopter/mode.cpp [hunk 1]
```diff
@@ -405,33 +405,15 @@ void Mode::get_pilot_desired_lean_angles(float &roll_out_cd, float &pitch_out_cd
...
-    // fetch roll and pitch stick positions
-    float thrust_angle_x_cd = - channel_pitch->get_control_in();
-    float thrust_angle_y_cd = channel_roll->get_control_in();
 
-    // limit max lean angle
-    angle_limit_cd = constrain_float(angle_limit_cd, 1000.0f, angle_max_cd);
+    //transform pilot's normalised roll or pitch stick input into a roll and pitch euler angle command
+    float roll_out_deg;
+    float pitch_out_deg;
+    rc_input_to_roll_pitch(roll_out_deg, pitch_out_deg, channel_roll->norm_input_dz(), channel_pitch->norm_input_dz(), angle_max_cd * 0.01,  angle_limit_cd * 0.01);
```
Comment (IamPete1) [inline ArduCopter/mode.cpp:412]: This change from `get_control_in` to `norm_input_dz` brakes simple and supper simple. 

You have to manually normalize the control in method for the time being.

**Review by IamPete1 — CHANGES_REQUESTED** [review]
Breaks simple mode. I guess we also need to update out simple mode auto test since it did not fail......

This PR will resolve the problem eventualy https://github.com/ArduPilot/ardupilot/pull/18581.

File: ArduCopter/mode.cpp (hunk 1 above)
Comment (lthall) [reply ArduCopter/mode.cpp]: Thanks Pete! Nice catch.

**Review by IamPete1 — APPROVED** [review]
LGTM

File: libraries/AP_Math/control.cpp [hunk 2]
```diff
@@ -508,18 +508,45 @@ float input_expo(float input, float expo)
...
+    float thrust_limit = tanf(radians(angle_limit_deg));
+
+    // apply horizontal thrust limit
+    thrust.limit_length(thrust_limit);
+
+    // Conversion from angular thrust vector to euler angles.
+    float pitch_rad = - atanf(thrust.x);
+    float roll_rad = atanf(cosf(pitch_rad) * thrust.y);
+
+    // Convert to centi-degrees
```
Comment (rmackay9) [inline libraries/AP_Math/control.cpp:549]: I guess this comment is out-of-date and it should be, "Convert to degrees"

File: libraries/AP_Math/control.cpp [hunk 3]
```diff
@@ -508,18 +508,45 @@ float input_expo(float input, float expo)
...
 {
     return degrees(atanf((accel/GRAVITY_MSS)));
 }
+
+// rc_input_to_roll_pitch - transform pilot's normalised roll or pitch stick input into a roll and pitch euler angle command
+// roll_in_unit and pitch_in_unit - are normalised roll and pitch stick input
+// angle_max_deg - maximum lean angle from the z axis
+// angle_limit_deg - provides the ability to reduce the maximum output lean angle to less than angle_max_deg
+// returns roll and pitch angle in degrees
+void rc_input_to_roll_pitch(float &roll_out_deg, float &pitch_out_deg, float roll_in_unit, float pitch_in_unit, float angle_max_deg, float angle_limit_deg)
```
Comment (rmackay9) [inline libraries/AP_Math/control.cpp:528]: It's a bit unconventional (at least in AP) to have the outputs as the first arguments. Normally they would be at the end I think.  @peterbarker do you have an opinion on this?

**Comment by rmackay9** [comment]
It is potentially scope creep but @tridge requested that the change be made to quadplanes as well.  We should decide if that is scope creep (and will be left to a future PR ) or not.

File: libraries/AP_Math/control.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_Math/control.cpp:528]: I concur - outputs last in ArduPilot (I've had to make the same change in other PRs).

Esp. so in this case as the method name is termed as "to" - having the arguments in the same order as in the method name is something I like.

**Comment by IamPete1** [comment]
@rmackay9 The quadplane code is a mess, I think its better to tackle that in a future PR.

File: libraries/AP_Math/control.cpp (hunk 2 above)
Comment (lthall) [reply libraries/AP_Math/control.cpp:549]: true

File: libraries/AP_Math/control.cpp (hunk 3 above)
Comment (lthall) [reply libraries/AP_Math/control.cpp:528]: Point well made! I will keep this in mind in the future!

**Comment by lthall** [comment]
I have made the change. Can somebody tell me how to change the last commit back to being authored by Mark?

**Review by tridge — APPROVED** [review]

**Pull request merged**